File: generic-runtime/nodejs/exclamation.js
async function process(message, context) {
    const logger = context.getLogger();
    for (let word of message.split(" ")) {
        await context.incrementCounter(word, 1)
        let count = await context.getCounter(word)
        logger.info(`got word: ${word} for ${count['value']} times`)
    }
    await context.publish("persistent://public/default/test-node-package-serde-extra", message.concat("!!!"))
    return message.concat("!");
}

File: generic-runtime/python/exclamation.py
from pulsar import Function

# The classic ExclamationFunction that appends an exclamation at the end
# of the input
class ExclamationFunction(Function):
    def __init__(self):
        pass

    def process(self, input, context):
        logger = context.get_logger()
        for word in input.split(" "):
            context.incr_counter(word, 1)
            count = context.get_counter(word)
            logger.info("got word: %s for %d times" % (word, count.value))
        context.publish("persistent://public/default/test-py-package-serde-extra", input + '!!!')
        return input + '!'